services/QuillLMS/app/models/student_response.rb

```diff
@@ -0,0 +1,257 @@
+#  fk_rails_...  (student_response_previous_feedback_text_id => student_response_previous_feedback_texts.id)
+#  fk_rails_...  (student_response_prompt_text_id => student_response_prompt_texts.id)
+#  fk_rails_...  (student_response_question_type_id => student_response_question_types.id)
+#
+class StudentResponse < ApplicationRecord
+  belongs_to :activity_session
+  belongs_to :question
+  belongs_to :student_response_answer_text
+  belongs_to :student_response_directions_text
+  belongs_to :student_response_instructions_text
+  belongs_to :student_response_previous_feedback_text
+  belongs_to :student_response_prompt_text
+  belongs_to :student_response_question_type
+
+  has_one :student_response_extra_metadata, dependent: :destroy
+
+  has_many :student_response_concept_results, dependent: :destroy
+  has_many :concept_results, through: :student_response_concept_results
+
+  has_many :student_responses_concepts, dependent: :destroy
+  has_many :concepts, through: :student_responses_concepts
+
+
+  validates_exclusion_of :correct, in: [nil]
+
+  # This is a list of keys from the old ConceptResults records
+  # that this model knows how to normalize.  Any keys that are
+  # fed into create_from_json that aren't from this list,
+  # will be stashed in a related StudentResponseExtraMetadata
+  # record
+  KNOWN_METADATA_KEYS = [
+    :answer,
+    :attemptNumber,
+    :correct,
+    :directions,
+    :lastFeedback,
+    :prompt,
+    :questionNumber,
+    :questionScore
+  ]
+
+  def self.create_from_json(data_hash)
+    data_hash = data_hash.deep_symbolize_keys
+
+    metadata = data_hash[:metadata]
+
+    question = calculate_question_from_hash(data_hash)
+
+    # Make sure that singular concept_id gets converted to array of
+    # conept_ids if that hasn't been done by an earlier step
+    data_hash[:concept_ids] = [data_hash[:concept_id]] unless data_hash.key?(:concept_ids) && !data_hash[:concept_ids].empty?
+
+    extra_metadata = extract_extra_metadata(metadata)
+
+    student_response = new(
+      activity_session_id: data_hash[:activity_session_id],
+      concept_ids: data_hash[:concept_ids].uniq,
+      attempt_number: metadata[:attemptNumber],
+      correct: metadata[:correct],
+      question: question,
+      question_number: metadata[:questionNumber],
+      question_score: metadata[:questionScore],
+      student_response_extra_metadata: extra_metadata
+    )
+
+    set_normalized_text(student_response, data_hash)
+
+    student_response.save!
+    student_response
+  end
+
+  def self.create_from_concept_result(concept_result)
+    return if concept_result.student_response
+
+    student_response = StudentResponse.find_by(
+      activity_session: concept_result.activity_session,
+      attempt_number: concept_result.metadata['attemptNumber'],
+      question_number: concept_result.metadata['questionNumber']
+    )
+
+    StudentResponse.transaction do
```

Comment:
I'd be curious with this transaction and the logic, how long with it take to run 1.7 billion of these?